support ranges of host ports to be mapped to a container port

brucekaufman opened the pull request:
this PR fixes a bug that causes starting localstack to fail to startup when specifying a port mapping from a range of host ports to a single container port EG `-p 9230-9231:9230`.  This turns out to be essential for being able to debug simultaneous nodejs lambda containers.  I am hoping to share the nodejs integration of localstack as described [here](https://github.com/localstack/localstack/issues/4443) but i discovered it was broken when trying to run on the newest version.

Specifying a host port range did work in version (0.12.16) but fails with the newest version of localstack.  To reproduce this you can pass a port range such as the example above into the `LAMBDA_DOCKER_FLAGS` variable of a localstack dockerfile.  

I have unit tested this (small) change but I was not able to build the docker image and confirm that the fix works because it looks like I encounter 4 failing integration tests when i try to build the docker container (before any of my changes are applied). I am happy to try to get this cleaned up and pushed through but will need guidance on how to build the container.

example error:
```
"errorMessage": "invalid literal for int() with base 10: '9230-9231'",
       "errorType": "ValueError",
       "stackTrace": Array [
         "  File \"/opt/code/localstack/localstack/services/awslambda/lambda_api.py\", line 811, in run_lambda
         lock_discriminator=lock_discriminator,
",
         "  File \"/opt/code/localstack/localstack/services/awslambda/lambda_executors.py\", line 429, in execute
         return do_execute()
     ",
         "  File \"/opt/code/localstack/localstack/services/awslambda/lambda_executors.py\", line 419, in do_execute
         return _run(func_arn=func_arn)
     ",
         "  File \"/opt/code/localstack/localstack/utils/cloudwatch/cloudwatch_util.py\", line 157, in wrapped
         raise e
     ",
         "  File \"/opt/code/localstack/localstack/utils/cloudwatch/cloudwatch_util.py\", line 153, in wrapped
         result = func(*args, **kwargs)
     ",
         "  File \"/opt/code/localstack/localstack/services/awslambda/lambda_executors.py\", line 406, in _run
         raise e
     ",
         "  File \"/opt/code/localstack/localstack/services/awslambda/lambda_executors.py\", line 402, in _run
         result = self._execute(lambda_function, inv_context)
     ",
         "  File \"/opt/code/localstack/localstack/services/awslambda/lambda_executors.py\", line 710, in _execute
         result = self.run_lambda_executor(lambda_function=lambda_function, inv_context=inv_context)
     ",
         "  File \"/opt/code/localstack/localstack/services/awslambda/lambda_executors.py\", line 599, in run_lambda_executor
         stdin=event_stdin_bytes,
     ",
         "  File \"/opt/code/localstack/localstack/services/awslambda/lambda_executors.py\", line 1119, in execute_in_container
         stdin=stdin,
     ",
         "  File \"/opt/code/localstack/localstack/utils/docker_utils.py\", line 1412, in run_container
         workdir=workdir,
     ",
         "  File \"/opt/code/localstack/localstack/utils/docker_utils.py\", line 1325, in create_container
         additional_flags, env_vars, ports, mount_volumes, network
     ",
         "  File \"/opt/code/localstack/localstack/utils/docker_utils.py\", line 953, in parse_additional_flags
         ports.add(int(host_port), int(container_port), protocol)
     ",
       ],
```

whummer commented:
Hi @brucekaufman - thanks for updating the PR! There's been quite a few changes in the meantime - can you please rebase your branch onto latest `master` and force-push the branch again? Thanks!

brucekaufman commented:
> Hi @brucekaufman - thanks for updating the PR! There's been quite a few changes in the meantime - can you please rebase your branch onto latest `master` and force-push the branch again? Thanks!

@whummer i've updated the PR and added handling for the corner case i discovered.  This is a fairly simple case but I'd still like to test this with the compiled docker image to 100% confirm my change is correct.  I'm embarrassed to say that when i ran `make docker-build` with the tests commented out, I couldn't figure out where the compiled image was getting output on the file system :(.  Can you point me in the right direction?  Other than that, the PR is good from my perspective.

whummer reviewed (commented):
Thanks for updating the PR @brucekaufman ! 👍 Added a few more minor comments. Apologies for the nitpicking, but this is a fairly central part of our logic, so it's important to ensure we have the different edge cases covered.. Thanks!

brucekaufman commented on localstack/utils/docker_utils.py:
```diff
@@ -142,6 +161,15 @@ def to_dict(self) -> Dict[str, Union[Tuple[str, int], int]]:
+                return [
+                    (
+                        f"{container_port}{protocol}",
+                        (bind_address, host_ports) if bind_address else host_ports,
```
Thank you, it has been a while since i worked with Python and the type annotation is new to me.

brucekaufman commented on localstack/utils/docker_utils.py:
```diff
@@ -174,6 +202,16 @@ def expand_range(self, port, range):
         else:
             raise Exception("Unable to add port %s to existing range %s" % (port, range))
 
+    def bisect_range(self, port, range):
```
I added something, please let me know if it makes sense.

brucekaufman commented on localstack/utils/docker_utils.py:
```diff
@@ -98,22 +98,39 @@ def add(self, port, mapped=None, protocol="tcp"):
+                bisected_host_port = from_range[0] + bisected_port_index
+                self.bisect_range(mapped, to_range)
+                self.bisect_range(bisected_host_port, from_range)
+                break
```
yes, `break` is the correct code here, although `continue` would still work, `break` is more efficient.  Since mappings cannot overlap, this code will only execute if the new mapping is contained within this one.  Therefore, it isn't necessary to run through the rest of the loop.

whummer replied:
Right, sorry, misread the logic in the loop - thanks for clarifying 👍

brucekaufman commented:
hi @whummer apologies for the delay here, i haven't had time to come to this until now.  I believe I have responded to your feedback, I also made some more changes to both the code and the test cases to make sure this is even more robust.  Please let me know I this is good.

whummer commented:
Thanks so much for updating the PR @brucekaufman - can I bother you one more time with fixing the lint issues please, see here: https://app.circleci.com/pipelines/github/localstack/localstack/2456/workflows/19008b78-ab33-4c7f-873a-73945ec6a92b/jobs/9354 (you can reproduce it locally by running `make lint`). Thanks!

dfangl commented on localstack/utils/docker_utils.py:
```diff
@@ -98,22 +98,50 @@ def add(self, port, mapped=None, protocol="tcp"):
+            else:
+                if not self.in_range(mapped, to_range):
+                    continue
+                # extending a 1 to 1 mapping to be many to 1
```
Is there a test for this behavior?
Like:
```
ports.add(5000:5000)
ports.add(5001:5000)
assert ports.to_str() = "-p 5000-5001:5000"
```
or something like that? Or do I miss the point of this lines?

brucekaufman replied:
@dfangl this case actually occurs when there is an adjacent port mapping to a many to one.  In this case it is in the extended range of the current mapping, but because a one to many map cannot be consolidated we continue the iteration.  This is covered by this case in the unit tests:
```
        port_mappings.add([7000, 7002], 7000)
        port_mappings.add(6999)
```

The example you've listed above actually is covered by adding any many to one mapping to the port mappings.  Because of the implementation where we split up a many to one and add each map recursively, the above example is actually equivalent to:

```
ports.add([5000, 5001], 5000)
```

In these cases we would not hit the if block above. Instead on each host-> container recurse we would simply hit this block each time we add a new mapping:
```
elif from_range_len == 1:
    self.expand_range(port, from_range)
```

dfangl replied:
This last block was actually the one I was talking about, should have put my comment on a range instead of a single line (the comment).
I agree the existing tests should cover that behavior.

dfangl commented on tests/unit/test_dockerclient.py:
```diff
@@ -111,3 +111,82 @@ def list_in(a, b):
     )
+
+
+class TestCommandLine(unittest.TestCase):
```
Nit: While fixing the Linter errors, could you rename this test suite into something more suiting? Like TestPortMappings or something like that. I know it has been named like this in the old location, but the name does not make sense here.

brucekaufman replied:
Done.

dfangl reviewed (commented):
In general, looks good to me :+1:

brucekaufman commented:
Sorry about missing the linter errors!  I had checked this but then ran into a merge conflict and forgot to run again.  If I do another PR in the future i'll get the pre-commit hooks configured.  I hope this is the last push! Also I have renamed the unit tests per @dfangl request

borg-z commented:
@dfangl @brucekaufman 
Now with the option `LAMBDA_DOCKER_FLAGS=-p19891-19895:19891` I get container with:
`0.0.0.0:19891->19891/tcp, 0.0.0.0:19892->19891/tcp, 0.0.0.0:19893->19891/tcp, 0.0.0.0:19894->19891/tcp, 0.0.0.0:19895->19891/tcp`
This is the wrong behavior. If you run two docker instances at the same time with  `docker run -it command --rm -p19891-19895:19891 alpine`, then you will see that docker selects free ports from the specified range.
This way I can run multiple containers at the same time with the same configuration without an error like "all ports are allocated."
![pic](https://user-images.githubusercontent.com/31611228/140924423-74c60fd6-98ad-40ba-bf7f-fca91b490506.png)

brucekaufman commented:
@dfangl @whummer the comment above from @borg-z is highlighting an issue.  Unfortunately it looks like while the behavior I implemented is 'correct' in the sense that it allows mapping multiple host ports to a single container port, it doesn't behave the way I would expect (and as this worked in the current version of localstack my team is pinned on).  Each container instance should not map ALL the host ports, it should simply select an open one from the provided range.  This allows multiple containers to be running simultaneously and still map the same container port to an open host port.

Looking at the documentation for the python SDK for docker, i'm not sure if it natively supports something like this or not.  I'm open to digging into this further but would like some guidance on the approach here.  Perhaps I should look at how localstack was working before the commit which broke this functionality?

dfangl commented:
Hi @brucekaufman !
The python sdk implementation is sadly quite undocumented, I had to dig into the source for quite some things.
In general, this SDK normally behaves like a thin wrapper around the docker engine API: https://docs.docker.com/engine/api/v1.41/

brucekaufman commented:
hi @dfangl I had time to dig into this some more today, and believe I have found the issue, but would like you to confirm before making the changes.  Based on your initial feedback to my PR, i assumed that we were using the `to_dict` method to pass the parameters to docker at runtime.  However, I only found one usage of the `PortMappings` class in the rest of the codebase, and it appears we use the to_list method instead, which I never tested or modified.  The reference is at line 616 in `localstack/utils/bootstrap.py`.  If you confirm this is correct this is a quick fix and I can put up a PR in short order.

With the example of @borg-z above, currently the `to_list` method returns the following, which I believe would result in all the ports being mapped rather than the expected behavior.

```
['-p', '19892-19893:19892-19891', '-p', '19892-19894:19892-19891', '-p', '19891-19895:19891-19891']
```
